src/jltypes.c

```diff
@@ -2476,27 +2492,67 @@ jl_datatype_t *jl_wrap_Type(jl_value_t *t)
         jl_instantiate_type_with((jl_value_t*)jl_type_type, env, 1);
 }
 
-void jl_reinstantiate_inner_types(jl_datatype_t *t)
+void jl_reinstantiate_inner_types(jl_datatype_t *t) // can throw!
 {
-    jl_typestack_t top;
+    inside_typedef = 0;
     assert(jl_is_datatype(t));
+    jl_typestack_t top;
     top.tt = t;
     top.prev = NULL;
-    size_t n = jl_svec_len(t->parameters);
-    if (n == 0) return;
-    t->name->cache = jl_emptysvec;
-    t->name->linearcache = jl_emptysvec;
-    jl_value_t **env = (jl_value_t**)alloca(n*2*sizeof(void*));
-    for(int i=0; i < n; i++) {
-        env[i*2] = jl_svecref(t->parameters,i);
-        env[i*2+1] = env[i*2];
+    size_t i, j, n = jl_svec_len(t->parameters);
+    if (n == 0) {
+        assert(partial_inst.len == 0);
+        return;
+    }
+
+    jl_value_t **env = (jl_value_t**)alloca(n * 2 * sizeof(void*));
+    for (i = 0; i < n; i++) {
+        env[i * 2] = jl_svecref(t->parameters, i);
+        env[i * 2 + 1] = NULL;
+    }
+
+    for (j = 0; j < partial_inst.len; j++) {
```

Comment:
I wanted to make all of the `super` arguments valid before working on the field arguments. It may not matter in the end, but it's not like there's any disadvantage to looping twice.